pkg/session/bootstrap.go

```diff
@@ -69,50 +69,51 @@ const (
-		Trigger_priv			ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Create_role_priv		ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Drop_role_priv			ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Account_locked			ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Shutdown_priv			ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Reload_priv				ENUM('N','Y') NOT NULL DEFAULT 'N',
-		FILE_priv				ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Config_priv				ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Create_Tablespace_Priv  ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Password_reuse_history  smallint unsigned DEFAULT NULL,
-		Password_reuse_time     smallint unsigned DEFAULT NULL,
-		User_attributes			json,
-		Token_issuer			VARCHAR(255),
-		Password_expired		ENUM('N','Y') NOT NULL DEFAULT 'N',
-		Password_last_changed	TIMESTAMP DEFAULT CURRENT_TIMESTAMP(),
-		Password_lifetime		SMALLINT UNSIGNED DEFAULT NULL,
+		Host					 CHAR(255),
+		User					 CHAR(32),
+		authentication_string	 TEXT,
+		plugin					 CHAR(64),
+		Select_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Insert_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Update_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Delete_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Drop_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Process_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Grant_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		References_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Alter_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Show_db_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Super_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_tmp_table_priv	 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Lock_tables_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Execute_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_view_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Show_view_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_routine_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Alter_routine_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Index_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_user_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Event_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Repl_slave_priv	    	 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Repl_client_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Trigger_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_role_priv		 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Drop_role_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Account_locked			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Shutdown_priv			 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Reload_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		FILE_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Config_priv				 ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Create_Tablespace_Priv   ENUM('N','Y') NOT NULL DEFAULT 'N',
+		Password_reuse_history   smallint unsigned DEFAULT NULL,
+		Password_reuse_time      smallint unsigned DEFAULT NULL,
+		Password_require_current ENUM('N','Y') DEFAULT NULL,
```

Comment:
how about put `Password_require_current` under column of `user_attirbutes` for better cross-version system table compatiblity?

Reply:
I prefer to put `Password_require_current` as a new field in `mysql.user` for better compatibility with mysql's 
https://dev.mysql.com/doc/mysql-security-excerpt/8.0/en/grant-tables.html#grant-tables-user-db

Reply:
it seems pr https://github.com/pingcap/tidb/pull/39460/ has addressed the compatibility problem here. so, i am fine with the change. but please make sure to do a cross version backup restore testing.

Reply:
Seems that we can add a new column to `mysql.user` with default value.
https://github.com/pingcap/tidb/pull/39460

Reply:
A backup of TiDB-with-this-PR to a vanilla TiDB returns this:
```
dvaneeden@dve-carbon:~$ tiup br restore full -s /tmp/b_full/
Checking updates for component br... Starting component br: /home/dvaneeden/.tiup/components/br/v8.3.0/br restore full -s /tmp/b_full/
Detail BR log in /tmp/br.log.2024-08-22T09.43.39+0200 
[2024/08/22 09:43:41.949 +02:00] [INFO] [collector.go:77] ["Full Restore failed summary"] [total-ranges=0] [ranges-succeed=0] [ranges-failed=0]
#######################################################################
# the target cluster is not compatible with the backup data,
# you can use '--with-sys-table=false' to skip restoring system tables
#######################################################################
Error: missing column in cluster data, table: user, col: Password_require_current enum('N','Y'): [BR:Restore:ErrRestoreIncompatibleSys]incompatible system table
```

Reply:
I can make BR to succeed with this patch:
```patch
diff --git a/br/pkg/restore/snap_client/systable_restore.go b/br/pkg/restore/snap_client/systable_restore.go
index a99d9925b1..7b758422d2 100644
--- a/br/pkg/restore/snap_client/systable_restore.go
+++ b/br/pkg/restore/snap_client/systable_restore.go
@@ -276,7 +276,7 @@ func (rc *SnapClient) replaceTemporaryTableToSystable(ctx context.Context, ti *m
                        zap.Stringer("schema", db.Name))
                // target column order may different with source cluster
                columnNames := make([]string, 0, len(ti.Columns))
-               for _, col := range ti.Columns {
+               for _, col := range db.ExistingTables[tableName].Columns {
                        columnNames = append(columnNames, utils.EncloseName(col.Name.L))
                }
                colListStr := strings.Join(columnNames, ",")
@@ -384,13 +384,19 @@ func CheckSysTableCompatibility(dom *domain.Domain, tables []*metautil.Table) er
                                col := backupTi.Columns[i]
                                clusterCol := clusterColMap[col.Name.L]
                                if clusterCol == nil {
-                                       log.Error("missing column in cluster data",
-                                               zap.Stringer("table", table.Info.Name),
-                                               zap.String("col", fmt.Sprintf("%s %s", col.Name, col.FieldType.String())))
-                                       return errors.Annotatef(berrors.ErrRestoreIncompatibleSys,
-                                               "missing column in cluster data, table: %s, col: %s %s",
-                                               table.Info.Name.O,
-                                               col.Name, col.FieldType.String())
+                                       if col.Name.L == "password_require_current" {
+                                               log.Error("ignoring missing column in cluster data",
+                                                       zap.Stringer("table", table.Info.Name),
+                                                       zap.String("col", fmt.Sprintf("%s %s", col.Name, col.FieldType.String())))
+                                       } else {
+                                               log.Error("missing column in cluster data",
+                                                       zap.Stringer("table", table.Info.Name),
+                                                       zap.String("col", fmt.Sprintf("%s %s", col.Name, col.FieldType.String())))
+                                               return errors.Annotatef(berrors.ErrRestoreIncompatibleSys,
+                                                       "missing column in cluster data, table: %s, col: %s %s",
+                                                       table.Info.Name.O,
+                                                       col.Name, col.FieldType.String())
+                                       }
                                }
                        }
                }
```

This patch:
- Makes the "missing column in cluster data" non-fatal for the "password_require_current" column
- Uses the list of columns for the target table instead of the list of columns of the source table. Maybe this should check for columns that are present in both if we actually want to do it like this.

However the result is that the `PASSWORD REQUIRE CURRENT (DEFAULT|OPTIONAL|)` part of the user definition gets lost as there isn't a column for that. I think this might be acceptable but not supporting this and requiring `--with-sys-table=false` might be a good option as well.

Reply:
And for the other way around there also seem to be isssues.

This is to restore a backup of v8.3.0 vanilla to TiDB-with-this-PR.

```
dvaneeden@dve-carbon:~/dev/pingcap/tidb$ ./bin/br restore full -s /tmp/b_full2
Detail BR log in /tmp/br.log.2024-08-22T10.35.17+0200 
Full Restore <----------------------------------------------------------------------------------------------------------------------------------------------------> 100.00%
[2024/08/22 10:35:21.503 +02:00] [INFO] [collector.go:77] ["Full Restore failed summary"] [total-ranges=16] [ranges-succeed=16] [ranges-failed=0] [split-region=194.686µs] [restore-ranges=7]
Error: error during merging temporary tables into system tables, table: user: [BR:Common:ErrUnknown]failed to execute REPLACE INTO `mysql`.`user`(`host`,`user`,`authentication_string`,`plugin`,`select_priv`,`insert_priv`,`update_priv`,`delete_priv`,`create_priv`,`drop_priv`,`process_priv`,`grant_priv`,`references_priv`,`alter_priv`,`show_db_priv`,`super_priv`,`create_tmp_table_priv`,`lock_tables_priv`,`execute_priv`,`create_view_priv`,`show_view_priv`,`create_routine_priv`,`alter_routine_priv`,`index_priv`,`create_user_priv`,`event_priv`,`repl_slave_priv`,`repl_client_priv`,`trigger_priv`,`create_role_priv`,`drop_role_priv`,`account_locked`,`shutdown_priv`,`reload_priv`,`file_priv`,`config_priv`,`create_tablespace_priv`,`password_reuse_history`,`password_reuse_time`,`password_require_current`,`user_attributes`,`token_issuer`,`password_expired`,`password_last_changed`,`password_lifetime`) SELECT `host`,`user`,`authentication_string`,`plugin`,`select_priv`,`insert_priv`,`update_priv`,`delete_priv`,`create_priv`,`drop_priv`,`process_priv`,`grant_priv`,`references_priv`,`alter_priv`,`show_db_priv`,`super_priv`,`create_tmp_table_priv`,`lock_tables_priv`,`execute_priv`,`create_view_priv`,`show_view_priv`,`create_routine_priv`,`alter_routine_priv`,`index_priv`,`create_user_priv`,`event_priv`,`repl_slave_priv`,`repl_client_priv`,`trigger_priv`,`create_role_priv`,`drop_role_priv`,`account_locked`,`shutdown_priv`,`reload_priv`,`file_priv`,`config_priv`,`create_tablespace_priv`,`password_reuse_history`,`password_reuse_time`,`password_require_current`,`user_attributes`,`token_issuer`,`password_expired`,`password_last_changed`,`password_lifetime` FROM `__tidb_br_temporary_mysql`.`user`;: [planner:1054]Unknown column 'password_require_current' in 'field list'
```

Reply:
Restoring a backup of TiDB v8.3.0 vanilla on TiDB-with-this-PR works:

From the logs:
```
[2024/08/22 10:49:10.696 +02:00] [WARN] [systable_restore.go:349] ["missing column in backup data"] [table=user] [col="Password_require_current enum('N','Y')"]
```